Revamp the way we build the PCH

nilsdeppe opened the pull request:
## Proposed changes

The way we used to build the PCH was very fragile and didn't work properly for
versions of CMake newer than 3.10. This completely revamps the way we build the
PCH with a detailed description of what we do and why.

### Types of changes:

- [ ] Bugfix
- [ ] New feature
- [x] Refactor

### Component:

- [ ] Code
- [ ] Documentation
- [x] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsdeppe committed:
Revamp the way we build the PCH

The way we used to build the PCH was very fragile and didn't work properly for
versions of CMake newer than 3.10. This completely revamps the way we build the
PCH with a detailed description of what we do and why.

nilsdeppe commented:
@kidder and @wthrowe would you mind checking to make sure this works correctly on your systems?

@wthrowe the bash is very cumbersome. When I tried things like `${flags[@]/delete_flag}`, GCC builds failed because they could find `` (empty file name?). I also try just setting `flags[i]=""`, which gave the same errors. If you have any suggestions on how to make it more elegant that would be very much appreciated!

kidder commented:
this works for me both on my macbook and my ubuntu desktop

wthrowe commented:
Haven't done a code review yet, just giving bash tips.

The usual way to handle this sort of thing is (untested)
```bash
pch_args=()
while [ $# -ne 0 ] ; do
  case "$1" in
    -c|-o)
      shift 2
      ;;
    *)
      pch_args+=("$1")
      shift
      ;;
  esac
done
```
Note that `$@` is empty at the end of this, so you'll need to save it in another array beforehand so you can run it at the end.  In the more common version of this pattern you would do something with `$2` before the `shift 2` statement.

This is hopefully pretty clear.  `$#` is the length of `$@`, `shift` removes the first thing in `$@` (so what was the second thing is now the first thing), and `shift 2` is the same as calling `shift` twice.  The `;;` is just the syntax for ending a case clause.  (There are other ones that do different things, but I can't remember what and I don't think I have ever seen anything except `;;` used.)

nilsdeppe commented:
I rebased on develop since the external lib changes required one minor update to this.

I also implemented @wthrowe's bash code (added in a fixup). This is much easier to read, thank you so much @wthrowe!

wthrowe commented:
Fails configure on a USE_PCH=NO build with lots of errors like
```
CMake Error at cmake/SpectreAddTestLibs.cmake:100 (add_dependencies):
  add_dependencies called with incorrect number of arguments
Call Stack (most recent call first):
  tests/Unit/ApparentHorizons/CMakeLists.txt:19 (add_test_library)
```

Do we still have a no-PCH CI build?  I saw some failures in that configuration earlier this week.

(Still haven't actually reviewed the code.)

nilsdeppe commented:
Rebased on develop and got it working with `USE_PCH=OFF`. Does the PCH work on your machine @wthrowe? If it doesn't I'd rather not keeping rebasing and maintaining this...

wthrowe commented:
Yes, my PCH builds passed before.  Just the one I test without failed.

nilsdeppe commented:
Okay, great, thank you for checking! And also thank you for testing the `USE_PCH=OFF`. I forgot to answer your question: no we no longer test `USE_PCH=OFF` on develop at all. It used to be only the IWYU build. I can see how long builds take. If it takes too long to build without the PCH we could configure CMake and make sure the non-test libs at executables build. While not completely thorough, it at least would cover most issues we might encounter. Probably just a debug clang-9 build is enough (since clang+debug builds fastest). What do you think?

wthrowe commented on tools/WrapPchCompiler.sh:
```diff
@@ -3,24 +3,21 @@
-        flags_to_use+=("${flags[$i]}")
-        continue
-    fi
+compiler_invokation="$@"
```
Use an array for this (`compiler_invokation=("$@")` and then `"${compiler_invokation[@]}"` later).  It handles quoting better.

wthrowe reviewed (changes requested):
Squash directly.

I've started my tests running.  They've all passed the configure stage, and all the ones that got that far last time were all fine, so you're probably good.

Yes, I think a single partial build would be enough for a CI test.  That will catch anything dramatic.  Any other problems should just be missing includes, and those are easy to fix later.

wthrowe commented:
Oh, and I forgot to say: Very nice documentation.

nilsdeppe commented:
Okay, I'll add a new CI target in a separate PR, thanks!

I made the change to the bash script. Thanks for explaining that to me! I was considering the two options while writing it but wasn't sure which was better. Now I know :)

Thanks! :)

wthrowe reviewed (dismissed):
Tests finished.  Looks good.

kidder commented on cmake/SetupPch.cmake:
```diff
@@ -3,192 +3,236 @@
+  # ${SPECTRE_PCH_HEADER_SOURCE_PATH} into the build dir as
+  # ${SPECTRE_PCH_DEP_HEADER_PATH} and generate a source file
+  # ${SPECTRE_PCH_DEP_SOURCE_PATH} that includes ${SPECTRE_PCH_DEP_HEADER_PATH}
+  # and thus tracks the includes of it dependencies. If we were to include
```
it -> its

nilsdeppe replied:
Done :)

kidder reviewed (commented):
one typo to squash

kidder commented:
@wthrowe this had a typo fixed
